erasure-code: add clay codes

tchaikov opened the pull request:
see also #23964

tchaikov commented:
@mynaramana i updated Sage's PR at #23964 with following changes

- delete trailing spaces
- re-indent some lines to comply with our [CodingStyle](https://github.com/ceph/ceph/blob/master/CodingStyle)
- wrap some lines so they don't exceed 80 chars limit
- move some definitions of variables to where they are used for the first time to improve the readability
- initialize some member variable in-class for better readability
- add `std::` and `ceph::` to some types used, as we cannot assume `using namespace std` or `using namespace ceph`. actually, the plugin does not compile without this change.
- remove unneeded constructor of `ScalarMDS`. the compiler creates a default one for us.
- mark `ErasureCodeClay::directory` const, as it won't change after being initialized in the constructor.
- move the constructor of `ErasureCodeClay` into .cc file, so we don't need to include `buffer.h` in the header.
- reduce the indent level of an inner loop in `ErasureCodeClay::repair_one_lost_chunk()`, in hope to improve the readability.
- add spaces around operators.
- add space around `//` in comments.
- replace patterns like `foo.find(i) == foo.end()` with `foo.count(i) == 0`, to avoid creating the temporary iterator instance.
- use structured binding and range-based loop when appropriate for better readability.
- use `initializer_list` to initialise `std::set<>`.
- replace `assert()` with `ceph_assert()`, the latter won't be optimized out even if `NDEBUG` is defined.
- use `unique_ptr<>` in `ErasureCodePluginClay::factory()` to simplify the error handling
- remove `BUILDING_FOR_EMBEDDED` handling, as this setting has been removed. it was once used by rook.

could you help review these change? also, i think you might want to update the header of the source files, unless 

- you think the Author is correct,  and 
- you want to grant the copyright of these files to the listed entities.
- you think the time of the first publication of your work was 2014. but since the paper based on which the clay code plugin was submitted in Sep, 2016, and the timestamp of the commit is Sep, 2018, i doubt it.

mynaramana commented:
@tchaikov Thanks a lot for the  changes. 

The author name needs to be updated in some files and yes the year should be 2018.  I am not really sure about what to do at the copyright part though. Could you suggest what should I be doing there ? Is it ok if I ignore the copyright lines altogether ?

If I have to make to changes to the headers which branch do I do those changes on ? Do I wait until this PR is stabilized ?

mynaramana commented on src/erasure-code/clay/ErasureCodeClay.cc:
```diff
@@ -334,62 +341,63 @@ int ErasureCodeClay::minimum_to_repair(const set<int> &want_to_read,
+	break;
+      }
+      if (!minimum->count(chunk)) {
+	minimum->emplace(chunk, sub_chunk_ind);
```
space needed?

mynaramana commented on src/erasure-code/clay/ErasureCodeClay.cc:
```diff
@@ -334,62 +341,63 @@ int ErasureCodeClay::minimum_to_repair(const set<int> &want_to_read,
-      } else break;
+    for (auto chunk : available_chunks) {
+      if (minimum->size() >= (unsigned)d) {
+	break;
```
space needed?

tchaikov replied:
@mynaramana it's a tab. because we use a tab-width of 8 and github's web interface does not respect the editor variable or our CodingStyle, the indent does not look right. but if you checkout the code, and open it using VIM or Emacs. it should look sane.

mynaramana commented on src/erasure-code/clay/ErasureCodeClay.cc:
```diff
@@ -509,137 +517,134 @@ int ErasureCodeClay::repair_one_lost_chunk(map<int, bufferlist> &recovered_data,
+  }
+
+  for (int order = 1; ;order++) {
+    if (ordered_planes.find(order) == ordered_planes.end()) {
```
ordered_planes.count(order) == 0

mynaramana commented on src/erasure-code/clay/ErasureCodeClay.cc:
```diff
@@ -509,137 +517,134 @@ int ErasureCodeClay::repair_one_lost_chunk(map<int, bufferlist> &recovered_data,
+	      i3 = 2;
+	    }
+	    if (aloof_nodes.count(node_sw) > 0) {
+	      assert(repair_plane_to_ind.find(z) != repair_plane_to_ind.end());
```
change it to count here and the next line as well

tchaikov commented:
@mynaramana i updated the PR to address your comment and also updated the header to list you as the copyright owner. i believe that's the right way to go, because i think you are the author of this creative work. unless you are employed by your institution for this work or your institution has a policy that states differently.

mynaramana commented:
> @mynaramana i updated the PR to address your comment and also updated the header to list you as the copyright owner. i believe that's the right way to go, because i think you are the author of this creative work. unless you are employed by your institution for this work or your institution has a policy that states differently.
> @tchaikov I am currently employed with my institute (Indian Institute of Science). I will find the contact information and write back to you soon on what email id can be used in the copyright line.

tchaikov commented:
@mynaramana i squashed my change into yours and also reverted the change to replace `$<TARGET_OBJECTS:unit-main>` with `GTest::Main` , as we need to initialize `g_ceph_context` for reading from settings from `g_conf()`.

tchaikov commented:
- http://pulpito.ceph.com/kchai-2018-09-28_04:23:08-rados-wip-kefu-testing-2018-09-27-2137-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-09-29_03:02:13-rados-wip-kefu-testing-2018-09-27-2137-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-09-29_07:58:27-rados-wip-kefu-testing-2018-09-27-2137-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-09-29_10:18:51-rados-wip-kefu-testing-2018-09-27-2137-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-09-29_13:24:35-rados-wip-kefu-testing-2018-09-27-2137-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-09-29_15:58:15-rados-wip-kefu-testing-2018-09-27-2137-distro-basic-smithi/

failed tests are tracked by http://tracker.ceph.com/issues/36260

tchaikov commented:
@jdurgin @neha-ojha ping

tchaikov commented:
http://pulpito.ceph.com/kchai-2018-10-02_01:17:59-rados-wip-kefu-testing-2018-09-27-2137-distro-basic-smithi/
